compiler/rustc_middle/src/ty/relate.rs

```diff
@@ -641,6 +641,35 @@ pub fn super_relate_consts<'tcx, R: TypeRelation<'tcx>>(
                 ty: a.ty(),
             }));
         }
+        (ty::ConstKind::Abstract(ae), ty::ConstKind::Abstract(be)) => {
+            let r = relation;
+            // FIXME(julianknodt): is it possible to relate two consts which are not identical
+            // exprs? Should we care about that?
+            let expr = match (ae, be) {
+                (Expr::Binop(a_op, al, ar), Expr::Binop(b_op, bl, br)) if a_op == b_op => {
+                    Expr::Binop(a_op, r.consts(al, bl)?, r.consts(ar, br)?)
+                }
+                (Expr::UnOp(a_op, av), Expr::UnOp(b_op, bv)) if a_op == b_op => {
+                    Expr::UnOp(a_op, r.consts(av, bv)?)
+                }
+                (Expr::Cast(ak, av, at), Expr::Cast(bk, bv, bt)) if ak == bk => {
+                    Expr::Cast(ak, r.consts(av, bv)?, r.tys(at, bt)?)
+                }
+                (Expr::FunctionCall(af, aa), Expr::FunctionCall(bf, ba))
+                    if aa.len() == ba.len() =>
+                {
+                    let mut related_args = Vec::with_capacity(aa.len());
+                    for (a_arg, b_arg) in aa.iter().zip(ba.iter()) {
+                        related_args.push(r.consts(*a_arg, *b_arg)?);
+                    }
+                    let related_args = tcx.arena.alloc_slice(&related_args);
+                    Expr::FunctionCall(r.consts(af, bf)?, related_args)
```

Comment:
Specifically here I think this is technically unsound -- if the function inspects the padding of its argument, then 'equal' consts can produce different results.

Reply:
Would inspecting the padding be permitted in unsafe code only?

Reply:
Inspecting padding bytes needs some unsafe code somewhere, yes.

But anyway this issue is probably existing already before this PR. I have been trying to explore in in rustc interactions, but I don't know how to even trigger this "unification of abstract/symbolic constants" case. But let's continue this in https://github.com/rust-lang/project-const-generics/issues/20.

Reply:
yeah, we need to deal with https://github.com/rust-lang/project-const-generics/issues/20 for this to be sound. Going to write a comment explaining my current thoughts about this at some point in the (hopefully near) future.

For now, please change it so that we first relate the functions, and then the arguments.